Update ClspvReflection non-semantic instruction set

callumfare opened the pull request:
Add support to the clspv reflection non-semantic instruction set for:
* Printf builtin
* Device address push constants for module-scope constants and program-scope global variables
* NumArguments, Flags, and Attributes for the Kernel instruction

This contribution is being made by Codeplay on behalf of Samsung.

alan-baker commented on include/spirv/unified1/extinst.nonsemantic.clspvreflection.grammar.json:
```diff
@@ -347,6 +348,67 @@
+          "value" : "0x0"
+        },
+        {
+          "enumerant" : "FlagMayUsePrintf",
```
```suggestion
          "enumerant" : "MayUsePrintf",
```

alan-baker reviewed (changes requested):
This will also need regenerated based on the resolution of https://github.com/KhronosGroup/SPIRV-Registry/pull/178

alan-baker commented on include/spirv/unified1/extinst.nonsemantic.clspvreflection.grammar.json:
```diff
@@ -1,12 +1,13 @@
         { "kind" : "IdRef", "name" : "Kernel" },
-        { "kind" : "IdRef", "name" : "Name" }
+        { "kind" : "IdRef", "name" : "Name" },
+        { "kind" : "IdRef", "name" : "Flags", "quantifier" : "?" }
```
This is not listed as optional in the instruction set I think.

alan-baker commented on include/spirv/unified1/extinst.nonsemantic.clspvreflection.grammar.json:
```diff
@@ -1,12 +1,14 @@
         { "kind" : "IdRef", "name" : "Kernel" },
-        { "kind" : "IdRef", "name" : "Name" }
+        { "kind" : "IdRef", "name" : "Name" },
+        { "kind" : "IdRef", "name" : "NumArguments"},
+        { "kind" : "IdRef", "name" : "Flags" }
```
Sorry there was a disconnect between the two reviews. These should both be optional for backwards compatibility. The validation for this instruction should eventually be updated to reject this instruction of the number of operands is not 2 in version 4 and 4 in version 5.

alan-baker replied:
I think the statement about the missing operands in the instruction set covers the optionality aspect in the spec language enough.

callumfare commented:
I've addressed the review feedback. I've also added a change from @FMarno which adds another operand to support clGetKernelInfo with `CL_KERNEL_ATTRIBUTES`
